src/optimizer/rule/IndexScanRule.cpp

```diff
@@ -0,0 +1,542 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/rule/IndexScanRule.h"
+#include "common/expression/LabelAttributeExpression.h"
+#include "optimizer/OptGroup.h"
+#include "planner/PlanNode.h"
+#include "planner/Query.h"
+
+namespace nebula {
+namespace opt {
+std::unique_ptr<OptRule> IndexScanRule::kInstance =
+    std::unique_ptr<IndexScanRule>(new IndexScanRule());
+
+IndexScanRule::IndexScanRule() {
+    RuleSet::queryRules().addRule(this);
+}
+
+bool IndexScanRule::match(const OptGroupExpr *groupExpr) const {
+    return groupExpr->node()->kind() == PlanNode::Kind::kIndexScan;
+}
+
+Status IndexScanRule::transform(graph::QueryContext *qctx,
+                                const OptGroupExpr *groupExpr,
+                                TransformResult *result) const {
+    FilterItems items;
+    ScanKind kind;
+    auto filter = filterExpr(groupExpr);
+    if (filter == nullptr) {
+        return Status::SemanticError();
+    }
+    auto ret = analyzeExpression(filter.get(), &items, &kind);
+    NG_RETURN_IF_ERROR(ret);
+
+    IndexQueryCtx iqctx = std::make_unique<std::vector<IndexQueryContext>>();
+    ret = createIndexQueryCtx(iqctx, kind, items, qctx, groupExpr);
+    NG_RETURN_IF_ERROR(ret);
+
+    auto newIG = cloneIndexScan(qctx, groupExpr);
+    newIG->setQueryContext(std::move(iqctx));
+    auto newGroupExpr = OptGroupExpr::create(qctx, newIG, nullptr);
+    result->newGroupExprs.emplace_back(newGroupExpr);
```

Comment:
assign true to result->eraseAll for erasing old optGroupExprs

Reply:
Good point.